pkg/controller/registry/grpc/source_test.go

```diff
@@ -0,0 +1,119 @@
+//go:generate counterfeiter -o ../../../fakes/fake_registry_store.go ../../../../vendor/github.com/operator-framework/operator-registry/pkg/registry/interface.go Query
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/registry/resolver"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/fakes"
+	"github.com/operator-framework/operator-registry/pkg/api"
+	"github.com/operator-framework/operator-registry/pkg/registry"
+	opserver "github.com/operator-framework/operator-registry/pkg/server"
+)
+
+func server(store registry.Query, port int) (func(), func()) {
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		logrus.Fatalf("failed to listen: %v", err)
+	}
+	s := grpc.NewServer()
+
+	api.RegisterRegistryServer(s, opserver.NewRegistryServer(store))
+
+	serve := func() {
+		if err := s.Serve(lis); err != nil {
+			logrus.Fatalf("failed to serve: %v", err)
+		}
+	}
+
+	stop := func() {
+		s.Stop()
+	}
+
+	return serve, stop
+}
+
+type FakeSourceSyncer struct {
+	// using a map[int] to preserve order
+	History map[resolver.CatalogKey]map[int]connectivity.State
+
+	sync.Mutex
+	expectedEvents int
+	done           chan struct{}
+}
+
+func (f *FakeSourceSyncer) sync(state SourceState) {
+	f.Lock()
+	logrus.Info(state)
+	if f.History[state.Key] == nil {
+		f.History[state.Key] = map[int]connectivity.State{}
+	}
+	f.History[state.Key][len(f.History[state.Key])] = state.State
+	f.expectedEvents -= 1
+	if f.expectedEvents == 0 {
+		f.done <- struct{}{}
+	}
+	f.Unlock()
+}
+
+func NewFakeSourceSyncer(expectedEvents int) *FakeSourceSyncer {
+	return &FakeSourceSyncer{
+		History:        map[resolver.CatalogKey]map[int]connectivity.State{},
+		expectedEvents: expectedEvents,
+		done:           make(chan struct{}),
+	}
+}
+
+func TestConnectionEvents(t *testing.T) {
+	type testcase struct {
+		name            string
+		expectedHistory map[resolver.CatalogKey]map[int]connectivity.State
+	}
+
+	test := func(tt testcase) func(t *testing.T) {
+		return func(t *testing.T) {
+			store := &fakes.FakeQuery{}
+			serve, stop := server(store, 50051)
+			go serve()
+			defer stop()
+
+			syncer := NewFakeSourceSyncer(len(tt.expectedHistory[resolver.CatalogKey{Name: "test", Namespace: "test"}]))
+			sources := NewSourceStore(logrus.New(), syncer.sync)
+
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+			sources.Start(ctx)
+
+			_, err := sources.Add(resolver.CatalogKey{Name: "test", Namespace: "test"}, "localhost:50051")
```

Comment:
It doesn't look like this test is complete. Should this call `Add` for each key in `tt.expectedHistory`?